src/main/java/com/aws/greengrass/deployment/KernelUpdateDeploymentTask.java

```diff
@@ -38,49 +42,72 @@ public class KernelUpdateDeploymentTask implements DeploymentTask {
     private final Logger logger;
     private final Deployment deployment;
     private final ComponentManager componentManager;
+    private final ExecutorService executorService;
+    private final CompletableFuture<DeploymentResult> deploymentResultCompletableFuture;
 
     /**
      * Constructor for DefaultDeploymentTask.
      *
-     * @param kernel Kernel instance
-     * @param logger Logger instance
-     * @param deployment Deployment instance
+     * @param kernel           Kernel instance
+     * @param logger           Logger instance
+     * @param deployment       Deployment instance
      * @param componentManager ComponentManager instance
+     * @param executorService  ExecutorService instance
      */
     public KernelUpdateDeploymentTask(Kernel kernel, Logger logger, Deployment deployment,
-                                      ComponentManager componentManager) {
+                                      ComponentManager componentManager, ExecutorService executorService) {
         this.kernel = kernel;
         this.deployment = deployment;
         this.logger = logger.dfltKv(DEPLOYMENT_ID_LOG_KEY, deployment.getGreengrassDeploymentId());
         this.componentManager = componentManager;
+        this.executorService = executorService;
+        this.deploymentResultCompletableFuture = new CompletableFuture<>();
     }
 
     @SuppressWarnings({"PMD.AvoidDuplicateLiterals"})
     @Override
     public DeploymentResult call() {
+        executorService.execute(this::waitForServicesToStart);
+        DeploymentResult result;
+        try {
+            result = deploymentResultCompletableFuture.get();
+        } catch (InterruptedException | ExecutionException | CancellationException e) {
+            return null;
```

Comment:
why are you returning null here? is this result not being handled anywhere else? should you throw this exception out instead?

Reply:
If we want to throw the exception then it would only be the interrupted exception.

The other two will occur when a deployment is cancelled, because the exceptions are already handled when waiting for services to start